Funcky/Extensions/EnumerableExtensions/Materialize.cs

```diff
@@ -22,15 +23,64 @@ public static IEnumerable<TItem> Materialize<TItem>(this IEnumerable<TItem> sour
         /// <param name="source">The source sequence can be any <see cref="IEnumerable{TItem}" />.</param>
         /// <param name="materialize">A function which materializes a given sequence into a collection.</param>
         /// <returns>A collection of the enumerated items.</returns>
-        public static IEnumerable<TItem> Materialize<TItem, TMaterialization>(
+        public static IReadOnlyCollection<TItem> Materialize<TItem, TMaterialization>(
             this IEnumerable<TItem> source,
             Func<IEnumerable<TItem>, TMaterialization> materialize)
-            where TMaterialization : IEnumerable<TItem>
-            => source is IReadOnlyCollection<TItem> or ICollection<TItem>
-                ? source
-                : materialize(source);
+            where TMaterialization : IReadOnlyCollection<TItem>
+            => source switch
+            {
+                IReadOnlyCollection<TItem> readOnlyCollection => readOnlyCollection,
+                IList<TItem> list => new ListAsReadOnlyCollectionProxy<TItem>(list),
+                ICollection<TItem> collection => new CollectionAsReadOnlyCollectionProxy<TItem>(collection),
+                _ => materialize(source),
+            };
 
-        private static ImmutableList<TItem> DefaultMaterialization<TItem>(IEnumerable<TItem> source)
+        private static IReadOnlyCollection<TItem> DefaultMaterialization<TItem>(IEnumerable<TItem> source)
```

Comment:
I need this class internal to use these helpers to implement Materialize in Funcky.Async

Reply:
66b88e1e5313e5cd8e6ecc9111018d8fe612866b